Fabric.Authorization.AccessControl/src/app/services/global/services.service.ts

```diff
@@ -37,73 +37,167 @@ export class ServicesService {
+            requireAuthToken: true
         }
     ];
 
-    constructor(private http: HttpClient, private configService: ConfigService) {}
-
-    public initialize(): Observable<string> {
-        return this.configService.getDiscoveryServiceRoot().pipe(
-            mergeMap(discoveryServiceRoot => {
-                const url: string =
-                    `${discoveryServiceRoot}/Services?$filter=` +
-                    this.services.map(service => `ServiceName eq \'${service.name}\'`).join(' or ') +
-                    `&$select=ServiceUrl,Version,ServiceName`;
-                return this.http.get<OData.IArray<IDiscoveryService>>(url, {withCredentials: true});
-            }),
+    constructor(private http: HttpClient, private configService: ConfigService) { }
+
+    public initialize() {
+
+    }
+
+    public getIdentityAndAccessControlUrl(): Observable<string[]> {
+        return this.isOAuthAuthenticationEnabled.pipe(
+            switchMap(isEnabled => {
+              if(isEnabled) {
+                // if OAuth, we get identity and access control up front
+                return forkJoin(this.identityServiceEndpoint, this.accessControlEndpoint)
+              } else {
+                // If Windows Auth, we have discovery up front and have to build
+                // out the service urls from discovery
+                return this.buildServiceMaps().toPromise().then(identityUrl => {
+                        return this.accessControlEndpoint.toPromise().then(accessControlUrl => {
+                            return [identityUrl, accessControlUrl]
+                        })
+                    })
+              }
+            })
+        )
+    }
+
+    private baseDiscoveryServiceUrl: string;
+    get discoveryServiceEndpoint(): Observable<string> {
+        if (this.baseDiscoveryServiceUrl === undefined) {
+            // if discovery service url is pushed to the website by the window.location, then use that
+            // if we get identity service url instead, then visit this website, grab the discovery_uri from 
+            // identity instead.
+            return this.configService.getDiscoveryServiceRoot()
+                .pipe(switchMap(url => url === '' ? this.discoveryServiceUrlFromIdentity : of(url)),
+                    map(url => this.trimRightChar(url, '/')),
+                    tap(url => this.baseDiscoveryServiceUrl = url));
+        }
+
+        return of(this.baseDiscoveryServiceUrl);
+    }
+
+    get isOAuthAuthenticationEnabled(): Observable<boolean> {
+        return this.configService.getUseOAuthAuthentication().pipe(
+            switchMap(stringBool => stringBool === 'true' ? of(true) : of(false))
+        )
+    }
+
+    get discoveryServiceUrlFromIdentity(): Observable<string> {
+        return this.configService.getIdentityServiceRoot()
+            .pipe(map((url) => `${url}/.well-known/openid-configuration`),
+                switchMap((url) => this.http.get(url)),
```

Comment:
if this call fails, how do we handle the error?

Reply:
Like our code before, there is no retry logic.  While I dont use this as an excuse, instead we can build an interceptor that would do some retry logic if we want?

I dont know if there is a neat RxJS way to retry, but the interceptor seems like a better idea all around for all calls.

Reply:
I see, so we should check the response to make sure its 200, otherwise throw error

Reply:
Verified that there is an interceptor that will catch exceptions from the HTTP client and display them in a toast message.  This should be good then